File: java/src/main/java/ai/buble/sdk/generations/CreateGenerationRequest.java
package ai.buble.sdk.generations;

import ai.buble.sdk.UnsupportedGenerationFieldException;

import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Request body for asynchronous media generation.
 *
 * <p>Buble's public generation API uses a flat JSON object. Stable fields are
 * represented directly on this type; model-specific parameters should be added
 * with {@link Builder#param(String, Object)} and are serialized at the JSON root.</p>
 */
public final class CreateGenerationRequest {
    private static final Set<String> FORBIDDEN_FIELDS = Collections.unmodifiableSet(new LinkedHashSet<String>(Arrays.asList(
            "input",
            "options",
            "scene",
            "sub_mode_id",
            "subModeId",
            "provider",
            "mediaType",
            "media_type",
            "images",
            "image_input",
            "video_input",
            "audio_input"
    )));

    private final String model;
    private final String mode;
    private final String prompt;
    private final List<String> imageUrls;
    private final String startFrame;
    private final String endFrame;
    private final List<String> videoUrls;
    private final List<String> audioUrls;
    private final Boolean isPublic;
    private final Boolean copyProtected;
    private final Map<String, Object> params;

    private CreateGenerationRequest(Builder builder) {
        this.model = builder.model;
        this.mode = builder.mode;
        this.prompt = builder.prompt;
        this.imageUrls = builder.imageUrls;
        this.startFrame = builder.startFrame;
        this.endFrame = builder.endFrame;
        this.videoUrls = builder.videoUrls;
        this.audioUrls = builder.audioUrls;
        this.isPublic = builder.isPublic;
        this.copyProtected = builder.copyProtected;
        this.params = Collections.unmodifiableMap(new LinkedHashMap<String, Object>(builder.params));
    }

    public static Builder builder() {
        return new Builder();
    }

    public Map<String, Object> toRequestBody() {
        Map<String, Object> body = new LinkedHashMap<String, Object>();
        put(body, "model", model);
        put(body, "mode", mode);
        put(body, "prompt", prompt);
        put(body, "image_urls", imageUrls);
        put(body, "start_frame", startFrame);
        put(body, "end_frame", endFrame);
        put(body, "video_urls", videoUrls);
        put(body, "audio_urls", audioUrls);
        put(body, "is_public", isPublic);
        put(body, "copy_protected", copyProtected);
        for (Map.Entry<String, Object> entry : params.entrySet()) {
            if (entry.getValue() == null) {
                continue;
            }
            if (FORBIDDEN_FIELDS.contains(entry.getKey())) {
                throw new UnsupportedGenerationFieldException(entry.getKey());
            }
            body.put(entry.getKey(), entry.getValue());
        }
        for (String key : body.keySet()) {
            if (FORBIDDEN_FIELDS.contains(key)) {
                throw new UnsupportedGenerationFieldException(key);
            }
        }
        return body;
    }

    private static void put(Map<String, Object> body, String key, Object value) {
        if (value == null) {
            return;
        }
        if (value instanceof String && ((String) value).isEmpty()) {
            return;
        }
        if (value instanceof List && ((List<?>) value).isEmpty()) {
            return;
        }
        body.put(key, value);
    }

    public static final class Builder {
        private String model;
        private String mode;
        private String prompt;
        private List<String> imageUrls;
        private String startFrame;
        private String endFrame;
        private List<String> videoUrls;
        private List<String> audioUrls;
        private Boolean isPublic;
        private Boolean copyProtected;
        private final Map<String, Object> params = new LinkedHashMap<String, Object>();

        public Builder model(String model) { this.model = model; return this; }
        public Builder mode(String mode) { this.mode = mode; return this; }
        public Builder prompt(String prompt) { this.prompt = prompt; return this; }
        public Builder imageUrls(List<String> imageUrls) { this.imageUrls = imageUrls; return this; }
        public Builder startFrame(String startFrame) { this.startFrame = startFrame; return this; }
        public Builder endFrame(String endFrame) { this.endFrame = endFrame; return this; }
        public Builder videoUrls(List<String> videoUrls) { this.videoUrls = videoUrls; return this; }
        public Builder audioUrls(List<String> audioUrls) { this.audioUrls = audioUrls; return this; }
        public Builder isPublic(Boolean isPublic) { this.isPublic = isPublic; return this; }
        public Builder copyProtected(Boolean copyProtected) { this.copyProtected = copyProtected; return this; }

        public Builder param(String key, Object value) {
            if (FORBIDDEN_FIELDS.contains(key)) {
                throw new UnsupportedGenerationFieldException(key);
            }
            this.params.put(key, value);
            return this;
        }

        public Builder params(Map<String, ?> params) {
            if (params != null) {
                for (Map.Entry<String, ?> entry : params.entrySet()) {
                    param(entry.getKey(), entry.getValue());
                }
            }
            return this;
        }

        public CreateGenerationRequest build() {
            return new CreateGenerationRequest(this);
        }
    }
}

File: swift/Sources/Buble/CreateGenerationRequest.swift
/// Request body for creating an asynchronous media generation task.
public struct CreateGenerationRequest: Sendable {
    public var model: String
    public var mode: String?
    public var prompt: String?
    public var imageURLs: [String]
    public var startFrame: String?
    public var endFrame: String?
    public var videoURLs: [String]
    public var audioURLs: [String]
    public var isPublic: Bool?
    public var copyProtected: Bool?
    public var parameters: [String: JSONValue]

    public init(model: String) {
        self.model = model
        self.imageURLs = []
        self.videoURLs = []
        self.audioURLs = []
        self.parameters = [:]
    }

    public func mode(_ value: String) -> CreateGenerationRequest {
        var copy = self
        copy.mode = value
        return copy
    }

    public func prompt(_ value: String) -> CreateGenerationRequest {
        var copy = self
        copy.prompt = value
        return copy
    }

    public func imageURLs(_ value: [String]) -> CreateGenerationRequest {
        var copy = self
        copy.imageURLs = value
        return copy
    }

    public func startFrame(_ value: String) -> CreateGenerationRequest {
        var copy = self
        copy.startFrame = value
        return copy
    }

    public func endFrame(_ value: String) -> CreateGenerationRequest {
        var copy = self
        copy.endFrame = value
        return copy
    }

    public func videoURLs(_ value: [String]) -> CreateGenerationRequest {
        var copy = self
        copy.videoURLs = value
        return copy
    }

    public func audioURLs(_ value: [String]) -> CreateGenerationRequest {
        var copy = self
        copy.audioURLs = value
        return copy
    }

    public func isPublic(_ value: Bool) -> CreateGenerationRequest {
        var copy = self
        copy.isPublic = value
        return copy
    }

    public func copyProtected(_ value: Bool) -> CreateGenerationRequest {
        var copy = self
        copy.copyProtected = value
        return copy
    }

    public func param(_ key: String, _ value: JSONValue) throws -> CreateGenerationRequest {
        try Self.assertSupportedField(key)
        var copy = self
        copy.parameters[key] = value
        return copy
    }

    func body() throws -> JSONValue {
        var object: [String: JSONValue] = ["model": .string(model)]
        if let mode, !mode.isEmpty { object["mode"] = .string(mode) }
        if let prompt, !prompt.isEmpty { object["prompt"] = .string(prompt) }
        if !imageURLs.isEmpty { object["image_urls"] = .array(imageURLs.map { .string($0) }) }
        if let startFrame, !startFrame.isEmpty { object["start_frame"] = .string(startFrame) }
        if let endFrame, !endFrame.isEmpty { object["end_frame"] = .string(endFrame) }
        if !videoURLs.isEmpty { object["video_urls"] = .array(videoURLs.map { .string($0) }) }
        if !audioURLs.isEmpty { object["audio_urls"] = .array(audioURLs.map { .string($0) }) }
        if let isPublic { object["is_public"] = .bool(isPublic) }
        if let copyProtected { object["copy_protected"] = .bool(copyProtected) }

        for (key, value) in parameters {
            try Self.assertSupportedField(key)
            object[key] = value
        }
        for key in object.keys {
            try Self.assertSupportedField(key)
        }
        return .object(object)
    }

    static func assertSupportedField(_ field: String) throws {
        if forbiddenFields.contains(field) {
            throw BubleError.unsupportedGenerationField(field: field)
        }
    }

    private static let forbiddenFields: Set<String> = [
        "input",
        "options",
        "scene",
        "sub_mode_id",
        "subModeId",
        "provider",
        "mediaType",
        "media_type",
        "images",
        "image_input",
        "video_input",
        "audio_input"
    ]
}
